routing/bicycle_directions.cpp

```diff
@@ -222,39 +165,172 @@ Index::FeaturesLoaderGuard & BicycleDirectionsEngine::GetLoader(MwmSet::MwmId co
 
   pathSegment.m_highwayClass = highwayClass;
   pathSegment.m_isLink = ftypes::IsLinkChecker::Instance()(ft);
-
   ft.GetName(FeatureType::DEFAULT_LANG, pathSegment.m_name);
-
-  pathSegment.m_nodeId = uniNodeId;
   pathSegment.m_onRoundabout = ftypes::IsRoundAboutChecker::Instance()(ft);
-  pathSegment.m_path = path;
-  // @TODO(bykoianko) It's better to fill pathSegment.m_weight.
+}
+
+void BicycleDirectionsEngine::GetUniNodeIdAndAdjacentEdges(
+    IRoadGraph::TEdgeVector const & outgoingEdges, FeatureID const & inFeatureId,
+    uint32_t startSegId, uint32_t endSegId, bool inIsForward, UniNodeId & uniNodeId,
+    BicycleDirectionsEngine::AdjacentEdges & adjacentEdges)
+{
+  // Outgoing edge angle is not used for bicycle routing.
+  adjacentEdges.m_outgoingTurns.isCandidatesAngleValid = false;
+  adjacentEdges.m_outgoingTurns.candidates.reserve(outgoingEdges.size());
+  uniNodeId = UniNodeId(inFeatureId, startSegId, endSegId, inIsForward);
+
+  for (auto const & edge : outgoingEdges)
+  {
+    auto const & outFeatureId = edge.GetFeatureId();
+    // Checking if |edge| is a fake edge.
+    if (!outFeatureId.IsValid())
+      continue;
+
+    FeatureType ft;
+    GetLoader(outFeatureId.m_mwmId).GetFeatureByIndex(outFeatureId.m_index, ft);
+
+    auto const highwayClass = ftypes::GetHighwayClass(ft);
+    ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Error, ());
+    ASSERT_NOT_EQUAL(highwayClass, ftypes::HighwayClass::Undefined, ());
+    adjacentEdges.m_outgoingTurns.candidates.emplace_back(0.0 /* angle */, uniNodeId, highwayClass);
+  }
+}
+
+bool BicycleDirectionsEngine::IsJoint(IRoadGraph::TEdgeVector const & ingoingEdges,
+                                      IRoadGraph::TEdgeVector const & outgoingEdges,
+                                      Edge const & ingoingRouteEdge, Edge const & outgoingRouteEdge)
+{
+  if (ingoingRouteEdge.GetFeatureId() != outgoingRouteEdge.GetFeatureId())
+    return true;
+
+  FeatureID const & featureId = ingoingRouteEdge.GetFeatureId();
+  uint32_t const segOut = outgoingRouteEdge.GetSegId();
+  for (Edge const & e : ingoingEdges)
+  {
+    // It's necessary to compare segments for cases when |featureId| is a loop.
+    if (e.GetFeatureId() != featureId || abs(static_cast<int>(segOut - e.GetSegId())) != 1)
+      return true;
+  }
+
+  uint32_t const segIn = ingoingRouteEdge.GetSegId();
+  for (Edge const & e : outgoingEdges)
+  {
+    // It's necessary to compare segments for cases when |featureId| is a loop.
+    if (e.GetFeatureId() != featureId || abs(static_cast<int>(segIn - e.GetSegId())) != 1)
+      return true;
+  }
+  return false;
+}
+
+bool BicycleDirectionsEngine::GetSegment(FeatureID const & featureId, uint32_t segId,
+                                         bool isFeatureForward, Segment & segment)
+{
+  if (!m_numMwmIds)
+    return false;
+
+  if (!featureId.m_mwmId.IsAlive())
+  {
+    ASSERT(featureId.m_mwmId.IsAlive(), ("Feature:", featureId, "is not alive."));
```

Comment:
If you do not keep MwmHandle, it's possible that there're no corresponding mwm, so it's not an error, it's an expected situation.